Message Display: Add a popover for group mentions.

ericeslinger opened the pull request:
Fixes #8300

This adds a click handler to `.user-group-mention` which works in a
fashion that is quite similar to `.user-mention`. It generates and
displays a popover.

The popover has a list of members, their online status (if they are not
bots) or their bot status if they are bots (not sure if bots can be
members of usergroups, but I'm able to add one, so I thought it would be
worth supporting).

The popover's `UL` element has max-height and overflow-y atttributes so
large groups will grow a scrollbar.

shubhamdhama commented:
Thanks for working on this @ericeslinger. Can you post a GIF for the change to give a quick feedback/review or you may discuss it at #design?

ericeslinger commented:
Hi @shubhamdhama. Here's a current screenshot of the popover. I centered the bot icon based on feedback from #frontend, and I'll post this also in #design (new to the community, so not sure in which channel to post).

Note, I've updated the PR branch, with this slightly new styling (and also because the earlier PR failed a linter check; I apparently forgot to run the linter last night).

![screenshot from 2018-02-14 09-39-39](https://user-images.githubusercontent.com/255715/36219131-43bb3b8c-116b-11e8-86d9-978796ba5560.png)

shubhamdhama commented:
It is fine at #frontend, actually, I was following this issue so wanted to review changes and your changes are looking great, [here is the link to conversation](https://chat.zulip.org/#narrow/stream/frontend/subject/8300.20layout.20feedback/near/498139) for the future reviewers. Thanks.

ericeslinger commented:
I've amended my previous commit to convert `=>` function syntax to `function` syntax, which is what it looks like the CI was not happy about.

ericeslinger commented:
I've amended again to include some tests for the new handlebars files.

ericeslinger commented:
Okay, tests passing. I've updated the layout a bit to include a request from Tim that there be a link to the usergroup manager in the popup (below is a screenshot of a very large group with the link in it). I also tweaked the alignment of the online / bot indicators a little bit, and of course added test coverage for the created handlebars files. Let me know what you think.

![screenshot from 2018-02-14 14-09-53](https://user-images.githubusercontent.com/255715/36281034-188aa1dc-1251-11e8-8c5c-ef60a850642a.png)

cc: @timabbott @brockwhittaker

timabbott commented on static/js/popovers.js:
```diff
@@ -133,6 +133,68 @@ function show_user_info_popover(element, user, message) {
+}
+
+
+
```
These extra blank lines are unnecessary.

timabbott commented on static/js/popovers.js:
```diff
@@ -449,6 +511,20 @@ exports.register_click_handlers = function () {
+    });
+
+
+
```
Same here; we just do 1 space between blocks in this file.

timabbott commented on static/templates/user_group_info_popover_content.handlebars:
```diff
@@ -0,0 +1,30 @@
+    <li>
+        <a href="#organization/user-groups-admin">
+            <i class="icon-vector-cog"></i>
+            Manage User Groups
```
This should be tagged for translation; see zulip.readthedocs.io/en/latest/translating/index.html

timabbott commented on static/js/popovers.js:
```diff
@@ -449,6 +511,20 @@ exports.register_click_handlers = function () {
+        var id = $(this).attr('data-user-group-id');
+        if (id === '*') {
+            return;
+        }
```
Why do you have a check for `*` here?  

(For context, `*` is a special value used in the older "user mention" feature for "wildcard mentions" -- i.e. mentioning everyone in the stream).

timabbott commented on static/js/popovers.js:
```diff
@@ -133,6 +133,68 @@ function show_user_info_popover(element, user, message) {
+            } else if (ypos > popover_size) {
+                placement = "top";
+            }
+        }
```
This logic for computing `placement` should probably be extracted to a function rather than copied from the user popover code path.

timabbott commented on static/js/popovers.js:
```diff
@@ -133,6 +133,68 @@ function show_user_info_popover(element, user, message) {
+        if (group === undefined) {
+            // This is never supposed to happen, not even for deactivated
+            // users, so we'll need to debug this error if it occurs.
+            blueslip.error('Bad sender in message' + message.sender_id);
```
This error message needs an update

timabbott commented on static/js/popovers.js:
```diff
@@ -133,6 +133,68 @@ function show_user_info_popover(element, user, message) {
+                    user_last_seen_time_status: user_last_seen_time_status(p.user_id),
+                });
+            })
+            .sort(function (a, b) { return a.full_name.localeCompare(b.full_name); }),
```
This fairly complex fetching and sorting logic might be best extracted as a separate function so that we can eventually unit-test aspects of it (like the sort order).

ericeslinger commented:
Hi @timabbott. I updated the scroll region to use perfectScrollba, refactored the fetch-and-sort of group members into two composable functions (with some null protection thrown in for good measure), and refactored the popover position decision into a single function called in both places.

timabbott commented:
Merged as 786bda674cddff3b22194120dac4137663338ae8, after some small tweaks to the commit message and a small whitespace issue in the node tests that I didn't think worth another round-trip.  Thanks @ericeslinger!
